app/webpacker/controllers/bulk_form_controller.js

```diff
@@ -10,12 +13,37 @@ export default class BulkFormController extends Controller {
     for (const element of this.form.elements) {
       element.addEventListener("keyup", this.toggleModified.bind(this)); // instant response
       element.addEventListener("change", this.toggleModified.bind(this)); // just in case (eg right-click paste)
+
+      // Set up a tree of fields according to their associated record
+      const recordContainer = element.closest("[data-record-id]"); // The JS could be more efficient if this data was added to each element. But I didn't want to pollute the HTML too much.
+      const recordId = recordContainer && recordContainer.dataset.recordId;
+      if (recordId) {
+        this.recordElements[recordId] ||= [];
+        this.recordElements[recordId].push(element);
+      }
     }
   }
 
   toggleModified(e) {
     const element = e.target;
-    const changed = element.value != element.defaultValue;
-    element.classList.toggle("modified", changed);
+    const modified = element.value != element.defaultValue;
+    element.classList.toggle("modified", modified);
+
+    this.toggleFormModified();
+  }
+
+  toggleFormModified() {
+    // For each record, check if any fields are modified
+    const modifiedRecordCount = Object.keys(this.recordElements).filter((recordId) => {
+      return this.recordElements[recordId].some((element) => {
+        return element.value != element.defaultValue;
+      });
+    }).length;
```

Comment:
Instead of iterating over `Object.keys(this.recordElements)` and then use `recordId` through `this.recordElements[recordId]` when not iterate over `Object.values()` ? Something like:

```js
const modifiedRecordCount = Object.values(this.recordElements).filter((elements) =>
      elements.some((element) => element.classList.contains("modified"))
    ).length;
```

Reply:
That's great, thank you! Also nice to see the cleaner one-line function format. I'm finally learning the most up to date Javascript syntax.

I decided to keep the definition of "modified" based on the defaultValue, but moved this to a private function, which cleans things up even further, yey!